### ISSUE TITLE
Stripe SDK retry on java.net.ConnectException

### ISSUE THREAD

**Issue by totof3110** [stripe/stripe-java#1071]
Stripe's Java SDK has had a retry mechanism for a little while. Looking at the source, it'll retry on exceptions that are instances of `java.net.ConnectException`.
```
    // Retry on connection error.
    if ((exception != null)
        && (exception.getCause() != null)
        && (exception.getCause() instanceof ConnectException)) {
      return true;
    }
```
However, we occasionally see in our logs the following stack trace:
```
Caused by: java.net.SocketTimeoutException: connect timed out
	at java.net.PlainSocketImpl.socketConnect(Native Method) ~[na:1.8.0_201]
	at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:350) ~[na:1.8.0_201]
	at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:206) ~[na:1.8.0_201]
	at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:188) ~[na:1.8.0_201]
	at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392) ~[na:1.8.0_201]
	at java.net.Socket.connect(Socket.java:589) ~[na:1.8.0_201]
	at sun.security.ssl.SSLSocketImpl.connect(SSLSocketImpl.java:666) ~[na:1.8.0_201]
	at sun.net.NetworkClient.doConnect(NetworkClient.java:175) ~[na:1.8.0_201]
	at sun.net.www.http.HttpClient.openServer(HttpClient.java:463) ~[na:1.8.0_201]
	at sun.net.www.http.HttpClient.openServer(HttpClient.java:558) ~[na:1.8.0_201]
	at sun.net.www.protocol.https.HttpsClient.<init>(HttpsClient.java:264) ~[na:1.8.0_201]
	at sun.net.www.protocol.https.HttpsClient.New(HttpsClient.java:367) ~[na:1.8.0_201]
	at sun.net.www.protocol.https.AbstractDelegateHttpsURLConnection.getNewHttpClient(AbstractDelegateHttpsURLConnection.java:191) ~[na:1.8.0_201]
	at sun.net.www.protocol.http.HttpURLConnection.plainConnect0(HttpURLConnection.java:1156) ~[na:1.8.0_201]
	at sun.net.www.protocol.http.HttpURLConnection.plainConnect(HttpURLConnection.java:1050) ~[na:1.8.0_201]
	at sun.net.www.protocol.https.AbstractDelegateHttpsURLConnection.connect(AbstractDelegateHttpsURLConnection.java:177) ~[na:1.8.0_201]
	at sun.net.www.protocol.http.HttpURLConnection.getInputStream0(HttpURLConnection.java:1564) ~[na:1.8.0_201]
	at sun.net.www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1492) ~[na:1.8.0_201]
	at java.net.HttpURLConnection.getResponseCode(HttpURLConnection.java:480) ~[na:1.8.0_201]
	at sun.net.www.protocol.https.HttpsURLConnectionImpl.getResponseCode(HttpsURLConnectionImpl.java:347) ~[na:1.8.0_201]
	at com.stripe.net.HttpURLConnectionClient.request(HttpURLConnectionClient.java:37) ~[stripe-java-19.24.0.jar:19.24.0]
	... 84 common frames omitted
```
The exception here is a `java.net.SocketTimeoutException` so the retry logic is not triggered and the call fails.

Any idea what could be causing those `java.net.SocketTimeoutException`s? Should the HTTP client retry when encountering them?

**Comment by richardm-stripe** [comment]
My impression is that `SocketTimeoutException` occurs if a socket has successfully been established, but too much time has elapsed without any bytes being written through it. Whereas 'ConnectException' includes timeouts for the socket becoming established in the first place.

Both seem suitable for retries to me -- I've added that behavior in #1072

**Comment by totof3110** [comment]
Thanks @richardm-stripe !

**Comment by richardm-stripe** [comment]
Thanks @totof3110 for the suggestion, it's been released in [v19.37.0](https://github.com/stripe/stripe-java/releases/tag/v19.37.0)

**Comment by totof3110** [comment]
Thanks @richardg-stripe!
